ENH: Support for free-threading

AFAIK the main issues include:
- Bit of global state in the `pyx` files
  - The `_lazy_import` and `numpy` check
- Lots of hypothesis tests
  - Not really a problem but they don't seem to play well with `--require-gil-disabled`
- Coverage also doesn't seem to do well with the `nogil`

But tests aside this already builds with minor changes:

```diff
diff --git i/setup.py w/setup.py
index 633bfa3..94e0896 100644
--- i/setup.py
+++ w/setup.py
@@ -4,6 +4,9 @@ import versioneer
 import shutil
 import glob
 
+from Cython.Compiler.Version import version as cython_version
+from packaging.version import Version
+
 with open("README.md", "r") as fh:
     long_description = fh.read()
 
@@ -12,6 +15,8 @@ from Cython.Build import cythonize
 CYTHON_COVERAGE = os.environ.get("CYTHON_COVERAGE", False)
 define_macros = []
 compiler_directives = {}
+if Version(cython_version) >= Version("3.1.0a0"):
+    compiler_directives["freethreading_compatible"] = True
 if CYTHON_COVERAGE:
     print("CYTHON_COVERAGE is set. Enabling Cython coverage support.")
     define_macros.append(("CYTHON_TRACE_NOGIL", "1"))
```

The test failures are probably nothing which can be handled here:

<details>
<summary>Test failure</summary>

```bash
pytest --require-gil-disabled        
==================================== test session starts =====================================
platform linux -- Python 3.13.3, pytest-8.3.5, pluggy-1.5.0
project deps: numpy-2.3.0.dev0+git20250304.6611d55
rootdir: /home/rgoswami/Git/Github/Quansight/freeThreading/ndindex
configfile: pytest.ini
plugins: cov-6.1.1, freethreaded-0.1.0, hypothesis-6.131.7
collected 122 items                                                                          

ndindex/tests/test_array.py ....                                                       [  3%]
ndindex/tests/test_as_subindex.py F                                                    [  4%]
ndindex/tests/test_booleanarray.py .F^C

========================================== FAILURES ==========================================
________________________________ test_as_subindex_hypothesis _________________________________
  + Exception Group Traceback (most recent call last):
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pytest_freethreaded/plugin.py", line 65, in get_one_result
  |     return item.runtest()
  |            ~~~~~~~~~~~~^^
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/_pytest/python.py", line 1627, in runtest
  |     self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pluggy/_hooks.py", line 513, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |            ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pluggy/_manager.py", line 120, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |            ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pluggy/_callers.py", line 139, in _multicall
  |     raise exception.with_traceback(exception.__traceback__)
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pluggy/_callers.py", line 103, in _multicall
  |     res = hook_impl.function(*args)
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
  |     result = testfunction(**testargs)
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/ndindex/ndindex/tests/helpers.py", line 414, in test_as_subindex_hypothesis
  |     def inner(*args, **kwargs):
  |                ^^^
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 1779, in wrapped_test
  |     _raise_to_user(errors, state.settings, [], " in explicit examples")
  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 1474, in _raise_to_user
  |     raise the_error_hypothesis_found
  | ExceptionGroup: Hypothesis found 3 distinct failures in explicit examples. (3 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/contextlib.py", line 141, in __enter__
    |     return next(self.gen)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/entropy.py", line 203, in deterministic_PRNG
    |     seed_all()
    |     ~~~~~~~~^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/entropy.py", line 175, in seed_all
    |     for k, r in RANDOMS_TO_MANAGE.items():
    |                 ~~~~~~~~~~~~~~~~~~~~~~~^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/weakref.py", line 213, in items
    |     for k, wr in self.data.items():
    |                  ~~~~~~~~~~~~~~~^^
    | RuntimeError: dictionary changed size during iteration
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 503, in execute_explicit_examples
    |     execute_example()
    |     ~~~~~~~~~~~~~~~^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 1024, in execute_once
    |     result = self.test_runner(data, run)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 729, in default_executor
    |     return function(data)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 996, in run
    |     return test(*args, **kwargs)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/ndindex/ndindex/tests/helpers.py", line 414, in test_as_subindex_hypothesis
    |     def inner(*args, **kwargs):
    |                ^^^^^^^^^^^^^^^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 893, in test
    |     with unwrap_markers_from_group(), ensure_free_stackframes():
    |                                       ~~~~~~~~~~~~~~~~~~~~~~~^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/junkdrawer.py", line 314, in __exit__
    |     warnings.warn(
    |     ~~~~~~~~~~~~~^
    |         "The recursion limit will not be reset, since it was changed "
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |     ...<2 lines>...
    |         stacklevel=2,
    |         ^^^^^^^^^^^^^
    |     )
    |     ^
    | hypothesis.errors.HypothesisWarning: The recursion limit will not be reset, since it was changed from another thread or during execution of a test.
    | Falsifying explicit example: test_as_subindex_hypothesis(
    |     idx1=(),
    |     idx2=(None, array([], dtype=int64)),
    |     shape=0,
    | )
    +---------------- 3 ----------------
    | Traceback (most recent call last):
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 503, in execute_explicit_examples
    |     execute_example()
    |     ~~~~~~~~~~~~~~~^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 1024, in execute_once
    |     result = self.test_runner(data, run)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 729, in default_executor
    |     return function(data)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 996, in run
    |     return test(*args, **kwargs)
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/ndindex/ndindex/tests/helpers.py", line 414, in test_as_subindex_hypothesis
    |     def inner(*args, **kwargs):
    |                ^^^^^^^^^^^^^^^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py", line 893, in test
    |     with unwrap_markers_from_group(), ensure_free_stackframes():
    |                                       ~~~~~~~~~~~~~~~~~~~~~~~^^
    |   File "/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/junkdrawer.py", line 314, in __exit__
    |     warnings.warn(
    |     ~~~~~~~~~~~~~^
    |         "The recursion limit will not be reset, since it was changed "
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |     ...<2 lines>...
    |         stacklevel=2,
    |         ^^^^^^^^^^^^^
    |     )
    |     ^
    | hypothesis.errors.HypothesisWarning: The recursion limit will not be reset, since it was changed from another thread or during execution of a test.
    | Falsifying explicit example: test_as_subindex_hypothesis(
    |     idx1=(),
    |     idx2=IntegerArray(0),
    |     shape=1,
    | )
    +------------------------------------

The above exception was the direct cause of the following exception:

item = <Function test_as_subindex_hypothesis>

    @pytest.hookimpl()
    def pytest_runtest_call(item: pytest.Item):
        # Try item.runtest()
        config_threads = item.config.option.threads
        config_iterations = item.config.option.iterations
        freethreaded_mark = item.get_closest_marker(name="freethreaded")
    
        if freethreaded_mark:
            threads = freethreaded_mark.kwargs.get("threads", config_threads)
            iterations = freethreaded_mark.kwargs.get("iterations", config_iterations)
        else:
            iterations = config_iterations
            threads = config_threads
    
        logger.debug("Running test %s", item.name)
        executor = ThreadPoolExecutor(max_workers=threads)
        barrier = threading.Barrier(threads)
        last_round = iterations % threads
        last_barrier = threading.Barrier(last_round) if last_round else None
        results = list(
            executor.map(
                get_one_result,
                repeat(item, iterations),
                chain(
                    repeat(barrier, iterations - last_round),
                    repeat(last_barrier, last_round),
                ),
            )
        )
        exceptions = [r for r in results if isinstance(r, Exception)]
        if not exceptions:
            return results[0]
        if len(exceptions) == len(results):
            raise results[0]
>       raise ConcurrencyError(
            iterations=iterations, failures=len(exceptions), threads=threads
        ) from exceptions[0]
E       pytest_freethreaded.plugin.ConcurrencyError: 197 failures in 200 iterations across 10 threads

../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pytest_freethreaded/plugin.py:104: ConcurrencyError
------------------------------------ Captured stdout call ------------------------------------
You can add @seed(315351513734118522597324920607266483358) to this test or run pytest with --hypothesis-seed=315351513734118522597324920607266483358 to reproduce this failure.
You can add @seed(315351513734118522597324920607266483358) to this test or run pytest with --hypothesis-seed=315351513734118522597324920607266483358 to reproduce this failure.
You can add @seed(50920028982634714371838145091489850749) to this test or run pytest with --hypothesis-seed=50920028982634714371838145091489850749 to reproduce this failure.
You can add @seed(50920028982634714371838145091489850749) to this test or run pytest with --hypothesis-seed=50920028982634714371838145091489850749 to reproduce this failure.
You can add @seed(260077405648651170589925033444372439773) to this test or run pytest with --hypothesis-seed=260077405648651170589925033444372439773 to reproduce this failure.
You can add @seed(260077405648651170589925033444372439773) to this test or run pytest with --hypothesis-seed=260077405648651170589925033444372439773 to reproduce this failure.
You can add @seed(260077405648651170589925033444372439773) to this test or run pytest with --hypothesis-seed=260077405648651170589925033444372439773 to reproduce this failure.
You can add @seed(260077405648651170589925033444372439773) to this test or run pytest with --hypothesis-seed=260077405648651170589925033444372439773 to reproduce this failure.
You can add @seed(260077405648651170589925033444372439773) to this test or run pytest with --hypothesis-seed=260077405648651170589925033444372439773 to reproduce this failure.
You can add @seed(206506616413695268040107946992465232270) to this test or run pytest with --hypothesis-seed=206506616413695268040107946992465232270 to reproduce this failure.
You can add @seed(206506616413695268040107946992465232270) to this test or run pytest with --hypothesis-seed=206506616413695268040107946992465232270 to reproduce this failure.
You can add @seed(194580732043707762818035134877460040746) to this test or run pytest with --hypothesis-seed=194580732043707762818035134877460040746 to reproduce this failure.
You can add @seed(272896614802485696722166168663134372125) to this test or run pytest with --hypothesis-seed=272896614802485696722166168663134372125 to reproduce this failure.
You can add @seed(156174108487074730511541248488581865242) to this test or run pytest with --hypothesis-seed=156174108487074730511541248488581865242 to reproduce this failure.
You can add @seed(156174108487074730511541248488581865242) to this test or run pytest with --hypothesis-seed=156174108487074730511541248488581865242 to reproduce this failure.
You can add @seed(156174108487074730511541248488581865242) to this test or run pytest with --hypothesis-seed=156174108487074730511541248488581865242 to reproduce this failure.
You can add @seed(106594411225802775793675091303063501702) to this test or run pytest with --hypothesis-seed=106594411225802775793675091303063501702 to reproduce this failure.
You can add @seed(184312408700286846428668432737841818328) to this test or run pytest with --hypothesis-seed=184312408700286846428668432737841818328 to reproduce this failure.
You can add @seed(50600870042117985830516491715987117876) to this test or run pytest with --hypothesis-seed=50600870042117985830516491715987117876 to reproduce this failure.
You can add @seed(50600870042117985830516491715987117876) to this test or run pytest with --hypothesis-seed=50600870042117985830516491715987117876 to reproduce this failure.
You can add @seed(50600870042117985830516491715987117876) to this test or run pytest with --hypothesis-seed=50600870042117985830516491715987117876 to reproduce this failure.
________________________________ test_booleanarray_hypothesis ________________________________

    @given(boolean_arrays, short_shapes)
>   def test_booleanarray_hypothesis(idx, shape):

ndindex/tests/test_booleanarray.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:794: in run
    self._run()
../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:1356: in _run
    self.generate_new_examples()
../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:967: in generate_new_examples
    zero_data = self.cached_test_function((ChoiceTemplate("simplest", count=None),))
../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:463: in cached_test_function
    self.test_function(data)
../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:475: in test_function
    self.__stoppable_test_function(data)
../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:366: in __stoppable_test_function
    with ensure_free_stackframes():
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hypothesis.internal.conjecture.junkdrawer.ensure_free_stackframes object at 0x44a6c1702b0>
args = (None, None, None), kwargs = {}

    def __exit__(self, *args, **kwargs):
        if self.new_maxdepth == sys.getrecursionlimit():
            sys.setrecursionlimit(self.old_maxdepth)
        else:  # pragma: no cover
>           warnings.warn(
                "The recursion limit will not be reset, since it was changed "
                "from another thread or during execution of a test.",
                HypothesisWarning,
                stacklevel=2,
            )
E           hypothesis.errors.HypothesisWarning: The recursion limit will not be reset, since it was changed from another thread or during execution of a test.

../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/junkdrawer.py:314: HypothesisWarning

The above exception was the direct cause of the following exception:

item = <Function test_booleanarray_hypothesis>

    @pytest.hookimpl()
    def pytest_runtest_call(item: pytest.Item):
        # Try item.runtest()
        config_threads = item.config.option.threads
        config_iterations = item.config.option.iterations
        freethreaded_mark = item.get_closest_marker(name="freethreaded")
    
        if freethreaded_mark:
            threads = freethreaded_mark.kwargs.get("threads", config_threads)
            iterations = freethreaded_mark.kwargs.get("iterations", config_iterations)
        else:
            iterations = config_iterations
            threads = config_threads
    
        logger.debug("Running test %s", item.name)
        executor = ThreadPoolExecutor(max_workers=threads)
        barrier = threading.Barrier(threads)
        last_round = iterations % threads
        last_barrier = threading.Barrier(last_round) if last_round else None
        results = list(
            executor.map(
                get_one_result,
                repeat(item, iterations),
                chain(
                    repeat(barrier, iterations - last_round),
                    repeat(last_barrier, last_round),
                ),
            )
        )
        exceptions = [r for r in results if isinstance(r, Exception)]
        if not exceptions:
            return results[0]
        if len(exceptions) == len(results):
            raise results[0]
>       raise ConcurrencyError(
            iterations=iterations, failures=len(exceptions), threads=threads
        ) from exceptions[0]
E       pytest_freethreaded.plugin.ConcurrencyError: 195 failures in 200 iterations across 10 threads

../pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/pytest_freethreaded/plugin.py:104: ConcurrencyError
------------------------------------ Captured stdout call ------------------------------------
You can add @seed(81159225522450987227151294230966413486) to this test or run pytest with --hypothesis-seed=81159225522450987227151294230966413486 to reproduce this failure.
You can add @seed(81159225522450987227151294230966413486) to this test or run pytest with --hypothesis-seed=81159225522450987227151294230966413486 to reproduce this failure.
You can add @seed(81159225522450987227151294230966413486) to this test or run pytest with --hypothesis-seed=81159225522450987227151294230966413486 to reproduce this failure.
You can add @seed(81159225522450987227151294230966413486) to this test or run pytest with --hypothesis-seed=81159225522450987227151294230966413486 to reproduce this failure.
You can add @seed(81159225522450987227151294230966413486) to this test or run pytest with --hypothesis-seed=81159225522450987227151294230966413486 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(265051599499098476648640623610739367539) to this test or run pytest with --hypothesis-seed=265051599499098476648640623610739367539 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(195521427344889582882463378660049832879) to this test or run pytest with --hypothesis-seed=195521427344889582882463378660049832879 to reproduce this failure.
You can add @seed(291393262175234170811802393324829355904) to this test or run pytest with --hypothesis-seed=291393262175234170811802393324829355904 to reproduce this failure.You can add @seed(291393262175234170811802393324829355904) to this test or run pytest with --hypothesis-seed=291393262175234170811802393324829355904 to reproduce this failure.

You can add @seed(291393262175234170811802393324829355904) to this test or run pytest with --hypothesis-seed=291393262175234170811802393324829355904 to reproduce this failure.
You can add @seed(291393262175234170811802393324829355904) to this test or run pytest with --hypothesis-seed=291393262175234170811802393324829355904 to reproduce this failure.
You can add @seed(291393262175234170811802393324829355904) to this test or run pytest with --hypothesis-seed=291393262175234170811802393324829355904 to reproduce this failure.
You can add @seed(291393262175234170811802393324829355904) to this test or run pytest with --hypothesis-seed=291393262175234170811802393324829355904 to reproduce this failure.
You can add @seed(49092876208473541602838304741021077141) to this test or run pytest with --hypothesis-seed=49092876208473541602838304741021077141 to reproduce this failure.
You can add @seed(49092876208473541602838304741021077141) to this test or run pytest with --hypothesis-seed=49092876208473541602838304741021077141 to reproduce this failure.
You can add @seed(49092876208473541602838304741021077141) to this test or run pytest with --hypothesis-seed=49092876208473541602838304741021077141 to reproduce this failure.
You can add @seed(49092876208473541602838304741021077141) to this test or run pytest with --hypothesis-seed=49092876208473541602838304741021077141 to reproduce this failure.
You can add @seed(49092876208473541602838304741021077141) to this test or run pytest with --hypothesis-seed=49092876208473541602838304741021077141 to reproduce this failure.
You can add @seed(198934648653087228638741576495841098214) to this test or run pytest with --hypothesis-seed=198934648653087228638741576495841098214 to reproduce this failure.
You can add @seed(321540919742720946716064541759488537467) to this test or run pytest with --hypothesis-seed=321540919742720946716064541759488537467 to reproduce this failure.
You can add @seed(321540919742720946716064541759488537467) to this test or run pytest with --hypothesis-seed=321540919742720946716064541759488537467 to reproduce this failure.
You can add @seed(321540919742720946716064541759488537467) to this test or run pytest with --hypothesis-seed=321540919742720946716064541759488537467 to reproduce this failure.
You can add @seed(275984478034492220141616147424496893488) to this test or run pytest with --hypothesis-seed=275984478034492220141616147424496893488 to reproduce this failure.
You can add @seed(275984478034492220141616147424496893488) to this test or run pytest with --hypothesis-seed=275984478034492220141616147424496893488 to reproduce this failure.
You can add @seed(275984478034492220141616147424496893488) to this test or run pytest with --hypothesis-seed=275984478034492220141616147424496893488 to reproduce this failure.
You can add @seed(8187593238448233606773601854612334783) to this test or run pytest with --hypothesis-seed=8187593238448233606773601854612334783 to reproduce this failure.
You can add @seed(8187593238448233606773601854612334783) to this test or run pytest with --hypothesis-seed=8187593238448233606773601854612334783 to reproduce this failure.
You can add @seed(8187593238448233606773601854612334783) to this test or run pytest with --hypothesis-seed=8187593238448233606773601854612334783 to reproduce this failure.
You can add @seed(8187593238448233606773601854612334783) to this test or run pytest with --hypothesis-seed=8187593238448233606773601854612334783 to reproduce this failure.
You can add @seed(8187593238448233606773601854612334783) to this test or run pytest with --hypothesis-seed=8187593238448233606773601854612334783 to reproduce this failure.
You can add @seed(8187593238448233606773601854612334783) to this test or run pytest with --hypothesis-seed=8187593238448233606773601854612334783 to reproduce this failure.
You can add @seed(319500550018170936743619473550084958932) to this test or run pytest with --hypothesis-seed=319500550018170936743619473550084958932 to reproduce this failure.
You can add @seed(195437120175260456481906223446181112062) to this test or run pytest with --hypothesis-seed=195437120175260456481906223446181112062 to reproduce this failure.
You can add @seed(195437120175260456481906223446181112062) to this test or run pytest with --hypothesis-seed=195437120175260456481906223446181112062 to reproduce this failure.
You can add @seed(195437120175260456481906223446181112062) to this test or run pytest with --hypothesis-seed=195437120175260456481906223446181112062 to reproduce this failure.
You can add @seed(195437120175260456481906223446181112062) to this test or run pytest with --hypothesis-seed=195437120175260456481906223446181112062 to reproduce this failure.
You can add @seed(91021958593266420379164491744950618948) to this test or run pytest with --hypothesis-seed=91021958593266420379164491744950618948 to reproduce this failure.
You can add @seed(198605782666507586697831004776285460380) to this test or run pytest with --hypothesis-seed=198605782666507586697831004776285460380 to reproduce this failure.
You can add @seed(198605782666507586697831004776285460380) to this test or run pytest with --hypothesis-seed=198605782666507586697831004776285460380 to reproduce this failure.
You can add @seed(198605782666507586697831004776285460380) to this test or run pytest with --hypothesis-seed=198605782666507586697831004776285460380 to reproduce this failure.
You can add @seed(198605782666507586697831004776285460380) to this test or run pytest with --hypothesis-seed=198605782666507586697831004776285460380 to reproduce this failure.
You can add @seed(161312717379017579509613467292512535939) to this test or run pytest with --hypothesis-seed=161312717379017579509613467292512535939 to reproduce this failure.
You can add @seed(161312717379017579509613467292512535939) to this test or run pytest with --hypothesis-seed=161312717379017579509613467292512535939 to reproduce this failure.
You can add @seed(161312717379017579509613467292512535939) to this test or run pytest with --hypothesis-seed=161312717379017579509613467292512535939 to reproduce this failure.
You can add @seed(33797491127006914791202521256327333657) to this test or run pytest with --hypothesis-seed=33797491127006914791202521256327333657 to reproduce this failure.
You can add @seed(33797491127006914791202521256327333657) to this test or run pytest with --hypothesis-seed=33797491127006914791202521256327333657 to reproduce this failure.
You can add @seed(33797491127006914791202521256327333657) to this test or run pytest with --hypothesis-seed=33797491127006914791202521256327333657 to reproduce this failure.
You can add @seed(265155431730091602713581599798306361033) to this test or run pytest with --hypothesis-seed=265155431730091602713581599798306361033 to reproduce this failure.
You can add @seed(265155431730091602713581599798306361033) to this test or run pytest with --hypothesis-seed=265155431730091602713581599798306361033 to reproduce this failure.
You can add @seed(265155431730091602713581599798306361033) to this test or run pytest with --hypothesis-seed=265155431730091602713581599798306361033 to reproduce this failure.
You can add @seed(189379348246229715972512423177997525053) to this test or run pytest with --hypothesis-seed=189379348246229715972512423177997525053 to reproduce this failure.
You can add @seed(189379348246229715972512423177997525053) to this test or run pytest with --hypothesis-seed=189379348246229715972512423177997525053 to reproduce this failure.
You can add @seed(189379348246229715972512423177997525053) to this test or run pytest with --hypothesis-seed=189379348246229715972512423177997525053 to reproduce this failure.
You can add @seed(189379348246229715972512423177997525053) to this test or run pytest with --hypothesis-seed=189379348246229715972512423177997525053 to reproduce this failure.
You can add @seed(189379348246229715972512423177997525053) to this test or run pytest with --hypothesis-seed=189379348246229715972512423177997525053 to reproduce this failure.
You can add @seed(13960375138997708183159485383561083502) to this test or run pytest with --hypothesis-seed=13960375138997708183159485383561083502 to reproduce this failure.
You can add @seed(265613140040532933965604472535486599909) to this test or run pytest with --hypothesis-seed=265613140040532933965604472535486599909 to reproduce this failure.
====================================== warnings summary ======================================
ndindex/tests/test_booleanarray.py::test_booleanarray_hypothesis
ndindex/tests/test_booleanarray.py::test_booleanarray_hypothesis
  /home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/internal/conjecture/engine.py:366: HypothesisWarning: The recursion limit will not be reset, since it was changed from another thread or during execution of a test.
    with ensure_free_stackframes():

ndindex/tests/test_booleanarray.py::test_booleanarray_hypothesis
  /home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/site-packages/hypothesis/core.py:893: HypothesisWarning: The recursion limit will not be reset, since it was changed from another thread or during execution of a test.
    with unwrap_markers_from_group(), ensure_free_stackframes():

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================== short test summary info ===================================
FAILED ndindex/tests/test_as_subindex.py::test_as_subindex_hypothesis - pytest_freethreaded.plugin.ConcurrencyError: 197 failures in 200 iterations across 10 thr...
FAILED ndindex/tests/test_booleanarray.py::test_booleanarray_hypothesis - pytest_freethreaded.plugin.ConcurrencyError: 195 failures in 200 iterations across 10 thr...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/home/rgoswami/Git/Github/Quansight/freeThreading/pixi_fthr/.pixi/envs/ndindex/lib/python3.13t/threading.py:359: KeyboardInterrupt
```

</details>

<!-- INSERT YOUR THOUGHTS SPECIFIC TO THE PACKAGE HERE:
- Are there relevant threading-related issues or docs?
- If there are native libraries (C, C++, etc.), are they thread-safe? Does code
  in the package already drop the GIL?
- Is there relevant global state?
- Are there particular parts of the code base for which dedicated
  multi-threaded tests should be written?
-->

The standard TODOs for adding free-threading support are:

- [ ] Audit Python bindings and declare them free-threading compatible (xref https://py-free-threading.github.io/porting/#updating-extension-modules).
- [ ] Run the test suite with `pytest-run-parallel` to find potential issues, and fix them.
- [ ] Run the test suite under ThreadSanitizer. _If possible, depends on how many dependencies there are and if they run under TSan._
- [ ] Add `cp313t-*` to CI to build free-threading wheels.

For more details, please see the [suggested plan of attack in the py-free-threading guide](https://py-free-threading.github.io/porting/#suggested-plan-of-attack).

Note that this is the first time I've looked at this repo, so I might be
missing known issues or code that needs closer inspection. Any suggestions here
will be very useful.

Let me know if I ought to hold off for any reason.

--- a/.github/workflows/freethreading_tests.yml
+++ b/.github/workflows/freethreading_tests.yml
@@ -1,0 +1,31 @@
+name: Free Threading Tests
+on: [push, pull_request]
+jobs:
+  tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Install uv and set the python version
+        uses: astral-sh/setup-uv@v5
+        with:
+          version: "0.7.2"
+          python-version: 3.13t
+          enable-cache: false
+      - name: Set up Python
+        run: uv python install
+      - name: Install Dependencies
+        run: |
+          uv pip install numpy pytest-cov pytest-timeout pytest-durations pytest-run-parallel sympy hypothesis
+      - name: Disable Coverage # see gh-198
+        run: |
+          sed -i '/plugins = Cython.Coverage/d' .coveragerc
+          sed -i 's/addopts/\#addopts/' pytest.ini
+      - name: Editable build
+        run: |
+          uv pip install -v -e .
+      - name: Run Tests
+        run: |
+          set -xe
+          uv run --no-project python -We:invalid -We::SyntaxWarning -m compileall -f -q ndindex/
+          uv run --no-project python -m pytest --parallel-threads=2 --iterations=2 -v -s \
+                 --timeout=600 --durations=10 -m "not hypothesis and not thread_unsafe and not flakes"

--- a/.github/workflows/publish-package.yml
+++ b/.github/workflows/publish-package.yml
@@ -89,7 +89,7 @@ jobs:
         run: echo "SDIST_FILE=$(ls dist/*.tar.gz)" >> $GITHUB_ENV
 
       - name: Build wheels
-        uses: pypa/cibuildwheel@v2.21.0
+        uses: pypa/cibuildwheel@v2.23.3
         with:
           package-dir: ${{ env.SDIST_FILE }}
           output-dir: dist/

--- a/.gitignore
+++ b/.gitignore
@@ -150,3 +150,4 @@ dmypy.json
 # Cython
 ndindex/*.c
 ndindex/*.html
+ndindex/*.cpp

--- a/ndindex/_slice.pyx
+++ b/ndindex/_slice.pyx
@@ -8,20 +8,20 @@ cdef extern from "Python.h":
     bint PyBool_Check(object obj)
     int64_t PyLong_AsLongLong(object obj) except? -1
 
+# Module scope, so implicitly thread safe
+# xref: https://github.com/Quansight-Labs/ndindex/pull/199#discussion_r2072395630
 cdef bint _NUMPY_IMPORTED = False
 cdef type _NUMPY_BOOL = None
+global _NUMPY_IMPORTED, _NUMPY_BOOL
+if not _NUMPY_IMPORTED:
+    if 'numpy' in sys.modules:
+        _NUMPY_BOOL = sys.modules['numpy'].bool_
+    _NUMPY_IMPORTED = True
 
 cdef inline bint is_numpy_bool(object obj):
-    global _NUMPY_IMPORTED, _NUMPY_BOOL
-    if not _NUMPY_IMPORTED:
-        if 'numpy' in sys.modules:
-            _NUMPY_BOOL = sys.modules['numpy'].bool_
-        _NUMPY_IMPORTED = True
     return _NUMPY_BOOL is not None and isinstance(obj, _NUMPY_BOOL)
 
 cdef inline int64_t cy_operator_index(object idx) except? -1:
-    cdef object result
-
     if PyBool_Check(idx) or is_numpy_bool(idx):
         raise TypeError(f"'{type(idx).__name__}' object cannot be interpreted as an integer")
 

--- a/ndindex/_tuple.pyx
+++ b/ndindex/_tuple.pyx
@@ -1,3 +1,4 @@
+# distutils: language = c++
 import sys
 
 # Forward declarations
@@ -51,7 +52,7 @@ cdef object _broadcast_shapes, _BroadcastError
 # We cannot just add these imports to the top because of circular import
 # issues. We can put them inside the constructor, but then they create a big
 # performance bottleneck.
-cdef void _lazy_import():
+cdef void _lazy_import() noexcept with gil:
     global _ndindex, _ArrayIndex, _Integer, _Slice, _BooleanArray, _IntegerArray, _ellipsis, _Newaxis
     global _broadcast_shapes, _BroadcastError
 
@@ -70,9 +71,20 @@ cdef void _lazy_import():
         _broadcast_shapes = broadcast_shapes
         _BroadcastError = BroadcastError
 
+cdef bint CYTHON_FREE_THREADING
+
+IF CYTHON_FREE_THREADING:
+    from libcpp.mutex cimport once_flag, call_once
+    cdef once_flag _lazy_import_lock
+
+
 cdef int _is_boolean_scalar(object idx):
     cdef object BooleanArray
-    _lazy_import()
+    IF CYTHON_FREE_THREADING:
+        with nogil:
+            call_once(_lazy_import_lock, _lazy_import)
+    ELSE:
+        _lazy_import()
     BooleanArray = _BooleanArray
     return isinstance(idx, BooleanArray) and idx.shape == ()
 
@@ -92,7 +104,11 @@ cdef class _Tuple:
             int has_boolean_scalar = 0
             object arg, newarg
 
-        _lazy_import()
+        IF CYTHON_FREE_THREADING:
+            with nogil:
+                call_once(_lazy_import_lock, _lazy_import)
+        ELSE:
+            _lazy_import()
 
         if 'numpy' in sys.modules:
             ndarray = sys.modules['numpy'].ndarray

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,3 +1,6 @@
 [build-system]
 requires = ["setuptools", "cython", "versioneer"]
 build-backend = "setuptools.build_meta"
+[tool.cibuildwheel]
+enable = ["cpython-freethreading"]
+skip = "*t*_i686"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,13 @@
 [build-system]
-requires = ["setuptools", "cython", "versioneer"]
+requires = [
+    # Pin to an older setuptools version for PyPy only, due to pypa/distutils#283
+    "setuptools; platform_python_implementation != 'PyPy'",
+    "setuptools<72.2; platform_python_implementation == 'PyPy'",
+    "cython",
+    "versioneer"
+]
 build-backend = "setuptools.build_meta"
+
 [tool.cibuildwheel]
 enable = ["cpython-freethreading"]
 skip = "*t*_i686"

--- a/setup.py
+++ b/setup.py
@@ -4,6 +4,9 @@ import versioneer
 import shutil
 import glob
 
+from Cython.Compiler.Version import version as cython_version
+from packaging.version import Version
+
 with open("README.md", "r") as fh:
     long_description = fh.read()
 
@@ -11,12 +14,17 @@ from Cython.Build import cythonize
 
 CYTHON_COVERAGE = os.environ.get("CYTHON_COVERAGE", False)
 define_macros = []
-compiler_directives = {}
+compiler_directives = dict()
+compiler_tenv = dict()
 if CYTHON_COVERAGE:
     print("CYTHON_COVERAGE is set. Enabling Cython coverage support.")
     define_macros.append(("CYTHON_TRACE_NOGIL", "1"))
     compiler_directives["linetrace"] = True
-
+if Version(cython_version) >= Version("3.1.0"):
+    compiler_directives["freethreading_compatible"] = True
+    compiler_tenv["CYTHON_FREE_THREADING"] = True
+else:
+    compiler_tenv["CYTHON_FREE_THREADING"] = False
 ext_modules = cythonize([
     Extension(
         "ndindex._slice", ["ndindex/_slice.pyx"],
@@ -25,9 +33,11 @@ ext_modules = cythonize([
     Extension(
         "ndindex._tuple", ["ndindex/_tuple.pyx"],
         define_macros=define_macros,
-    )],
+    )
+],
     language_level="3",
     compiler_directives=compiler_directives,
+    compile_time_env=compiler_tenv,
 )
 
 class CleanCommand(Command):

